File: app/src/main/java/com/icostel/arhitecturesample/di/modules/ViewModelModule.java
package com.icostel.arhitecturesample.di.modules;


import com.icostel.arhitecturesample.di.ViewModelFactory;
import com.icostel.arhitecturesample.di.ViewModelKey;
import com.icostel.arhitecturesample.ui.loginuser.LoginUserViewModel;
import com.icostel.arhitecturesample.ui.listusers.ListUsersViewModel;
import com.icostel.arhitecturesample.ui.newuser.NewUserActivity;
import com.icostel.arhitecturesample.ui.newuser.NewUserViewModel;
import com.icostel.arhitecturesample.ui.userdetails.UserDetailsViewModel;
import com.icostel.arhitecturesample.utils.error.ErrorViewModel;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;

@Module
public abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(ListUsersViewModel.class)
    abstract ViewModel bindListUsersViewModel(ListUsersViewModel listUsersViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(ErrorViewModel.class)
    abstract ViewModel bindErrorViewModel(ErrorViewModel listUsersViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(LoginUserViewModel.class)
    abstract ViewModel bindLoginUserViewModel(LoginUserViewModel loginUserViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(UserDetailsViewModel.class)
    abstract ViewModel bindUserDetailsViewModel(UserDetailsViewModel userDetailsViewModel);

    @Binds
    @IntoMap
    @ViewModelKey(NewUserViewModel.class)
    abstract ViewModel bindNewUserViewModel(NewUserViewModel newUserViewModel);

    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);
}

File: app/src/main/java/com/icostel/arhitecturesample/di/modules/ViewModelModule.kt
package com.icostel.arhitecturesample.di.modules


import androidx.lifecycle.ViewModel
import com.icostel.arhitecturesample.di.keys.ViewModelKey
import com.icostel.arhitecturesample.ui.listusers.ListUsersViewModel
import com.icostel.arhitecturesample.ui.loginuser.LoginUserViewModel
import com.icostel.arhitecturesample.ui.newuser.NewUserViewModel
import com.icostel.arhitecturesample.ui.splashscreen.SplashScreenViewModel
import com.icostel.arhitecturesample.ui.userdetails.UserDetailsViewModel
import com.icostel.arhitecturesample.utils.error.ErrorViewModel
import dagger.Binds
import dagger.Module
import dagger.multibindings.IntoMap

@Module
abstract class ViewModelModule {

    @Binds
    @IntoMap
    @ViewModelKey(ListUsersViewModel::class)
    abstract fun bindListUsersViewModel(listUsersViewModel: ListUsersViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(ErrorViewModel::class)
    abstract fun bindErrorViewModel(listUsersViewModel: ErrorViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(LoginUserViewModel::class)
    abstract fun bindLoginUserViewModel(loginUserViewModel: LoginUserViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(UserDetailsViewModel::class)
    abstract fun bindUserDetailsViewModel(userDetailsViewModel: UserDetailsViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(NewUserViewModel::class)
    abstract fun bindNewUserViewModel(newUserViewModel: NewUserViewModel): ViewModel

    @Binds
    @IntoMap
    @ViewModelKey(SplashScreenViewModel::class)
    abstract fun bindSplashScreenViewModel(splashScreenViewModel: SplashScreenViewModel): ViewModel
}
